### PR TITLE
 not defining properties until the connectedCallback for Elements

### PR DESCRIPTION
**Pull request by phillipskevin**
For https://github.com/canjs/can-define-class/issues/31.

### CODE REVIEW

File: can-define-class.js [hunk 1]
```diff
@@ -88,15 +90,39 @@ function define(Base = Object) {
...
+		connectedCallback() {
+			if (typeof super.connectedCallback === "function") {
+				super.connectedCallback();
+			}
+			this.setup( this[constructorPropsSymbol] );
+		}
+
+		setup(props) {
+			if(!this[hasBeenSetupSymbol]) {
+				Object.defineProperty(this, inSetupSymbol, {
```
Comment (matthewp) [inline can-define-class.js]: I think there are a few defineProperty() calls in this file now, what do you think about adding a helper function so we can call `setNonEnumerable(obj, symbol, value)` or something?

File: test/define-test.js [hunk 2]
```diff
@@ -17,3 +17,85 @@ QUnit.test("Can define stuff", function(assert) {
   let faves = new Faves();
   assert.equal(faves.color, "blue", "Got the value");
 });
+
+QUnit.test("Can not define stuff", function(assert) {
```
Comment (matthewp) [inline test/define-test.js]: What is this test testing?

File: test/define-test.js [hunk 3]
```diff
@@ -17,3 +17,85 @@ QUnit.test("Can define stuff", function(assert) {
...
+	constructor() {
+		super();
+		assert.equal(this.color, "blue", "color exists after constructor");
+	}
+  }
+
+  let faves = new Faves();
+});
+
+QUnit.test("Stuff is defined in connectedCallback for custom elements", function(assert) {
```
Comment (matthewp) [inline test/define-test.js]: Saucelabs is failing in Edge, I'm guessing for a lack of custom element support.  I would detect CE support and only run these test in supported browsers.

Comment (matthewp) [reply test/define-test.js]: `'customElements' in window` should be enough.

**Comment by matthewp** [comment]
How does this work with the lifecycle stuff's initialize hook? Should this be implementing initialize?

File: test/define-test.js (hunk 2 above)
Comment (phillipskevin) [reply test/define-test.js]: It's testing that it doesn't throw if you done have a `static define`.

File: test/define-test.js (hunk 3 above)
Comment (phillipskevin) [reply test/define-test.js]: Should we just add the polyfill? That's what I'm doing in `can-stache-define-element`.

**Comment by phillipskevin** [comment]
The lifecycle stuff will call the `setup` function in the `initialize` hook. It basically looks like this right now:

```js
module.exports = function mixinDefineClass(Base = HTMLElement) {
        return class Defined extends dFine(Base) {
                construct(props) {
                        super.construct(props);
                        this[constructorPropsSymbol] = props;
                        this.constructor._initDefines();
                }

                initialize() {
                        super.initialize();
                        this.setup.call(this, this[constructorPropsSymbol], false);
                }
        };
};
```

File: test/define-test.js (hunk 2 above)
Comment (matthewp) [reply test/define-test.js]: There's an existing test for that: https://github.com/canjs/can-define-class/blob/bf83944ef3499f64e9e01974a30e468af8505be9/test/define-class-test.js#L143-L146

File: test/define-test.js (hunk 3 above)
Comment (matthewp) [reply test/define-test.js]: I don't think its appropriate in this case, can-define-class is not about creating custom elements. This is just testing that it works correctly when used with custom elements.

File: can-define-class.js [hunk 4]
```diff
@@ -62,13 +64,22 @@ function getKeyValue(key) {
 	}
 }
 
+function defineNonEnumerable(obj, prop, value) {
```
Comment (matthewp) [inline can-define-class.js]: Is this any different from the one exported by define.js? If not can we just use that one? https://github.com/canjs/can-define-class/blob/bf83944ef3499f64e9e01974a30e468af8505be9/define.js#L1145

File: test/define-test.js (hunk 2 above)
Comment (phillipskevin) [reply test/define-test.js]: Ok, I'll look into what the difference is. In my case, `this.define` is undefined here: https://github.com/canjs/can-define-class/blob/bf83944ef3499f64e9e01974a30e468af8505be9/can-define-class.js#L71.

**Pull request merged**